app/src/common/shared/org/mozilla/vrbrowser/browser/adapter/ComponentsAdapter.kt

```diff
@@ -44,18 +64,88 @@ class ComponentsAdapter private constructor(
         ))
     }
 
-    fun link(tabId: String, geckoSession: GeckoSession) {
+    fun link(session: Session) {
         store.dispatch(EngineAction.LinkEngineSessionAction(
-                tabId,
-                GeckoEngineSession(geckoSession)
+                session.id,
+                GeckoEngineSession(session)
         ))
     }
 
-    fun unlink(tabId: String) {
+    fun unlink(session: Session) {
+        store.state.extensions.map { it.value }.forEach {
+            if (it.popupSessionId == session.id) {
+                store.dispatch(
+                        WebExtensionAction.UpdatePopupSessionAction(it.id, popupSession = null)
+                )
+            }
+        }
         store.dispatch(EngineAction.UnlinkEngineSessionAction(
-                tabId
+                session.id
+        ))
+    }
+
+    fun updateUrl(sessionId: String, url: String) {
+        store.dispatch(ContentAction.UpdateUrlAction(
+                sessionId, url
         ))
     }
+
+    init {
+        store.flowScoped { flow ->
+            flow.ifChanged { it.extensions }
+                    .map { it.extensions.filterValues { extension -> extension.popupSession != null } }
+                    .ifChanged()
+                    .collect { extensionStates ->
+                        if (extensionStates.values.isNotEmpty()) {
+                            val state = store.state.tabs.firstOrNull { tabState ->
+                                extensionStates.values.any { extState ->
+                                    val popUpGeckoSession = (extState.popupSession as GeckoEngineSession).geckoSession
+                                    val tabGeckoSession = (tabState.engineState.engineSession as GeckoEngineSession).geckoSession
+                                    popUpGeckoSession == tabGeckoSession
```

Comment:
Here you're looking for the tab that displays the action popup, right? You can also just replace this with `store.state.tabs.find { it.engineState.engineSession === popupSession }`

Reply:
One part I don't understand here is that you're passing `openPopupInTab = true` to `WebExtensionSupport`. So, you want all browser/page actions to be displayed in a separate tab (instead of a popup). Basically, what Fennec does which is fine, but then the `extension.popupSession` is never set and you'd have to find the tab by ID, using `extension.popupSessionId`. Happy to pair on this / discuss this more.